modules/kvformat/filterx-func-parse-kv.c

```diff
@@ -0,0 +1,384 @@
+
+      ok = _set_json_value(result, name, value, value_len);
+      if (!ok)
+        break;
+
+      if (self->stray_words_value_name)
+        {
+          gsize stray_words_value_len = kv_scanner_get_stray_words_len(&self->scanner);
+          if ( stray_words_value_len > 0)
+            ok = _set_json_value(result, self->stray_words_value_name, kv_scanner_get_stray_words(&self->scanner),
+                                 stray_words_value_len);
+        }
+      if (!ok)
+        break;
+    }
+
+exit:
+  if (!ok)
+    {
+      filterx_object_unref(result);
+    }
+  filterx_object_unref(obj);
+  return ok?result:NULL;
+}
+
+static void
+_free(FilterXExpr *s)
+{
+  FilterXFunctionParseKV *self = (FilterXFunctionParseKV *) s;
+  filterx_expr_unref(self->msg);
+  g_free(self->prefix);
+  g_free(self->pair_separator);
+  g_free(self->stray_words_value_name);
+  kv_scanner_deinit(&self->scanner);
+  filterx_function_free_method(&self->super);
+}
+
+static FilterXExpr *
+_extract_parse_kv_msg_expr(GList *argument_expressions, GError **error)
+{
+  FilterXExpr *msg_expr = filterx_expr_ref(((FilterXExpr *) argument_expressions->data));
+  if (!msg_expr)
+    {
+      g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                  "argument must be set: msg_str. " FILTERX_FUNC_PARSE_KV_USAGE);
+      return NULL;
+    }
+
+  return msg_expr;
+}
+
+static GPtrArray *
+_extract_parse_kv_opts(GList *argument_expressions, GError **error)
+{
+  gsize arguments_len = argument_expressions ? g_list_length(argument_expressions) : 0;
+  if (arguments_len < 1)
+    {
+      g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                  "invalid number of arguments. " FILTERX_FUNC_PARSE_KV_USAGE);
+      return NULL;
+    }
+
+  GPtrArray *opts = g_ptr_array_new_full(arguments_len - 1, g_free);
+
+  guint32 i = 0;
+  for (GList *elem = argument_expressions->next; elem; elem = elem->next)
+    {
+      if (i >= PARSE_KV_OPTS_COUNT)
+        break;
+
+      FilterXExpr *argument_expr = (FilterXExpr *) elem->data;
+
+      if (!argument_expr || !filterx_expr_is_literal(argument_expr))
+        {
+          g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                      "%s argument must be string literal. " FILTERX_FUNC_PARSE_KV_USAGE,
+                      parse_kv_opts_names[i]);
+          goto error;
+        }
+
+      FilterXObject *format_obj = filterx_expr_eval(argument_expr);
+      if (!format_obj)
+        {
+          g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                      "%s argument. " FILTERX_FUNC_PARSE_KV_USAGE, parse_kv_opts_names[i]);
+          goto error;
+        }
+
+      if (filterx_object_is_type(format_obj, &FILTERX_TYPE_NAME(string)))
+        {
+          gsize format_len;
+          const gchar *opt = filterx_string_get_value(format_obj, &format_len);
```

Comment:
[optional]

I think this is a recurring theme. Maybe we could have a `filterx_expr_extract_string_literal()` or sth similar.